presto-hive/src/main/java/com/facebook/presto/hive/StoragePartitionLoader.java

```diff
@@ -351,53 +360,77 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
+        ListMultimap<Integer, HiveFileInfo> bucketToFileInfo = ArrayListMultimap.create();
 
         if (!shouldCreateFilesForMissingBuckets(table, session)) {
             fileInfos.stream()
-                    .forEach(fileInfo -> bucketToFileInfo.put(getBucketNumber(fileInfo.getPath().getName()), fileInfo));
+                    .forEach(fileInfo -> {
+                        String fileName = fileInfo.getPath().getName();
+                        OptionalInt bucket = getBucketNumber(fileName);
+                        if (bucket.isPresent()) {
+                            bucketToFileInfo.put(bucket.getAsInt(), fileInfo);
+                        }
+                        else {
+                            throw new PrestoException(HIVE_INVALID_BUCKET_FILES, format("invalid hive bucket file name: %s", fileName));
+                        }
+                    });
         }
         else {
-            // verify we found one file per bucket
-            if (fileInfos.size() != partitionBucketCount) {
-                throw new PrestoException(
-                        HIVE_INVALID_BUCKET_FILES,
-                        format("Hive table '%s' is corrupt. The number of files in the directory (%s) does not match the declared bucket count (%s) for partition: %s",
-                                new SchemaTableName(table.getDatabaseName(), table.getTableName()),
-                                fileInfos.size(),
-                                partitionBucketCount,
-                                partitionName));
-            }
-
-            if (fileInfos.get(0).getPath().getName().matches("\\d+")) {
-                try {
-                    // File names are integer if they are created when file_renaming_enabled is set to true
-                    fileInfos.sort(Comparator.comparingInt(fileInfo -> Integer.parseInt(fileInfo.getPath().getName())));
+            // build mapping of file name to bucket
+            for (HiveFileInfo file : fileInfos) {
+                String fileName = file.getPath().getName();
+                OptionalInt bucket = getBucketNumber(fileName);
+                if (bucket.isPresent()) {
+                    bucketToFileInfo.put(bucket.getAsInt(), file);
+                    continue;
                 }
-                catch (NumberFormatException e) {
+
+                // legacy mode requires exactly one file per bucket
+                if (fileInfos.size() != partitionBucketCount) {
                     throw new PrestoException(
-                            HIVE_INVALID_FILE_NAMES,
-                            format("Hive table '%s' is corrupt. Some of the filenames in the partition: %s are not integers",
-                                    new SchemaTableName(table.getDatabaseName(), table.getTableName()),
+                            HIVE_INVALID_BUCKET_FILES,
+                            format("Hive table '%s' is corrupt. File '%s' does not match the standard naming pattern, and the number " +
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: &#39;size&#39; expression, partitionBucketCount (declared at line 348). This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 396. Consider formatting this data with &#39;%d&#39; instead.